SAML Implementation in Kratos

ThibHrrd opened the pull request:
# SAML Implementation 

## Related issue(s)

https://github.com/ory/kratos/issues/1928

psauvage0 commented on embedx/config.schema.json:
```diff
@@ -312,6 +312,225 @@
+            "Microsoft Active Directory"
+          ]
+        },
+        "public_cert_path": {
```
certificate_path ? Or public_key_certificate_path ?

psauvage0 commented on embedx/config.schema.json:
```diff
@@ -312,6 +312,225 @@
+            }
+          ]
+        },
+        "attributes_map": {
```
The attributes mapping is given in the config ? What if I have an attribute that is not provided here ? For example `group` ?

ThibHrrd replied:
For the moment, this part is not configurable, that is to say that the identity cannot have more fields than what is present here.

psauvage0 commented on internal/httpclient/api/openapi.yaml:
```diff
@@ -1185,6 +1185,51 @@ paths:
       summary: Create a Logout URL for Browsers
       tags:
       - v0alpha2
+  /self-service/methods/saml/auth:
```
Not related to the doc but to the route of this call (I don't know where the route is defined). 

In the [Ory Documentation for the API](https://www.ory.sh/docs/kratos/reference/api#operation/initializeSelfServiceLoginFlowWithoutBrowser), we can see that when there are different calls to make for browser vs API, they usually end the route with `/browser` or `/api`.  Maybe the same can be applied here ?

sebferrer replied:
We did it on purpose based on a misunderstanding, indeed we should change it back

psauvage0 commented on selfservice/flow/saml/handler.go:
```diff
@@ -0,0 +1,325 @@
+	}
+
+	// Key pair to encrypt and sign SAML requests
+	keyPair, err := tls.LoadX509KeyPair(strings.Replace(c.SAMLProviders[len(c.SAMLProviders)-1].PublicCertPath, "file://", "", 1), strings.Replace(c.SAMLProviders[len(c.SAMLProviders)-1].PrivateKeyPath, "file://", "", 1))
```
No Out Of Bound possible here ?

psauvage0 commented on selfservice/flow/saml/handler.go:
```diff
@@ -0,0 +1,325 @@
+	var idpMetadata *samlidp.EntityDescriptor
+
+	// We check if the metadata file is provided
+	if c.SAMLProviders[len(c.SAMLProviders)-1].IDPInformation["idp_metadata_url"] != "" {
```
OOB ? 
Also, you can check if a map entry is set with something like `if value, ok := mymap["key"]; ok { ...`
```suggestion
	if metadataURL, ok := c.SAMLProviders[len(c.SAMLProviders)-1].IDPInformation["idp_metadata_url"]; ok {
```

Stoakes replied:
I feel like this access lacks safety railings:

 * if len(c.SAMLProviders) == 0 this method will panic. Example at https://go.dev/play/p/AAD95wCtX_P
 * Is IDPInformation always initialized ?
 * Is idp_metadata_url always set ? Aren't there cases when it is empty but still valid ?

```suggestion
	if len(c.SAMLProviders) > 0 {
	     if val, ok := c.SAMLProviders[len(c.SAMLProviders)-1].IDPInformation["idp_metadata_url"]; ok && len(val) > 0 {
```

psauvage0 commented on selfservice/flow/saml/handler.go:
```diff
@@ -0,0 +1,325 @@
+	// Sometimes there is an issue with double slash into the url so we prevent it
+	// Crewjam library use default route for ACS and metadat but we want to overwrite them
+	RouteSamlAcsWithSlash := RouteSamlAcs
+	if publicUrlString[len(publicUrlString)-1] != '/' {
```
You could probably simplify this with something like `strings.TrimSuffix(strings.TrimSuffix(publicURLString, "/"), "/") + "/"`.
See https://pkg.go.dev/strings#TrimSuffix for more info

psauvage0 commented on selfservice/flow/saml/handler.go:
```diff
@@ -0,0 +1,325 @@
+			return err
+		}
+		samlMiddleWare.ServiceProvider.AcsURL = *u
+	}
```
```suggestion
  u, err := url.Parse(strings.TrimSuffix(strings.TrimSuffix(publicURLString, "/"), "/") + "/"+ RouteSamlAcsWithSlash)
  if err != nil {
	  return err
  }
  samlMiddleWare.ServiceProvider.AcsURL = *u

```

psauvage0 commented on selfservice/flow/saml/handler.go:
```diff
@@ -0,0 +1,325 @@
+
+func GetMiddleware() (*samlsp.Middleware, error) {
+	if samlMiddleware == nil {
+		return nil, errors.Errorf("The MiddleWare for SAML is null (Probably due to a backward step)")
```
There is no null in go :p

psauvage0 replied:
Joke aside, I don't understand the backward step part, the error message is a bit unclear

psauvage0 commented on selfservice/strategy/saml/provider.go:
```diff
@@ -0,0 +1,41 @@
+	Config() *Configuration
+}
+
+type Claims struct {
```
I couldn't find anything like this type Claims in the other strategies... Is it used to handle attribute mapping ? If so see previous comment about attribute mapping config

psauvage0 replied:
I couldn't find Claims like those in other strategies. Is it linked to attribute mapping ?

psauvage0 commented on selfservice/strategy/saml/provider_saml.go:
```diff
@@ -0,0 +1,62 @@
+		return nil, errors.Wrapf(err, "Unable to decode config %v", string(conf))
+	}
+
+	claims := &Claims{
+		Issuer:        "saml",
+		Subject:       SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["id"]),
+		Name:          SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["firstname"]),
+		LastName:      SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["lastname"]),
+		Nickname:      SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["nickname"]),
+		Gender:        SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["gender"]),
+		Birthdate:     SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["birthdate"]),
+		Picture:       SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["picture"]),
+		Email:         SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["email"]),
+		PhoneNumber:   SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["phone_number"]),
+		EmailVerified: true,
+	}
```
Attribute mapping needs to be able to take into account user defined attributes. What if I need the attribute `group` from the IDP to map to a `group` trait I configured in a Kratos Identity schema ?

ThibHrrd replied:
For the moment, the user can't add attributes as he wants but maybe I should add a group attribute

psauvage0 commented on selfservice/strategy/saml/strategy/strategy.go:
```diff
@@ -0,0 +1,352 @@
+
+// Handle /selfservice/methods/saml/acs | Receive SAML response, parse the attributes and start auth flow
+func (s *Strategy) handleCallback(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	r.ParseForm()
```
Error is not checked

psauvage0 commented on selfservice/strategy/saml/strategy/strategy.go:
```diff
@@ -0,0 +1,352 @@
+		return nil, err
+	}
+
+	provider, err := c.Provider(c.SAMLProviders[len(c.SAMLProviders)-1].ID, c.SAMLProviders[len(c.SAMLProviders)-1].Label)
```
No OOB possible ?

Stoakes commented on .schema/api.openapi.json:
```diff
@@ -1932,6 +1932,81 @@
     },
+    "/self-service/saml/metadata":{
+      "get":{
+        "description": "This endpoint is for the IDP to obtain kratos metadata",
```
The style for the other descriptions is slightly different.

"This endpoints returns kratos metadata as part of a SAMl login" would be closer to what is done on other endpoints.

Stoakes commented on .schema/api.openapi.json:
```diff
@@ -1932,6 +1932,81 @@
+    },
+    "/self-service/saml/idp":{
+      "get":{
+        "description": "This endpoint is to redirect the user to the idp auth flow",
```
Same as above: style for description text is different from other already existing endpoints.

psauvage0 commented on selfservice/strategy/saml/strategy/strategy.go:
```diff
@@ -0,0 +1,352 @@
+				return err
+			}
+
+			traitNodes, err := container.NodesFromJSONSchema(r.Context(), node.OpenIDConnectGroup, ds.String(), "", nil)
```
Use SAMLGroup node ?

Stoakes commented on .schema/api.openapi.json:
```diff
@@ -1932,6 +1932,81 @@
+      "get":{
+        "description": "This endpoint is for the IDP to obtain kratos metadata",
+        "operationId": "getSamlMetadata",
+        "response": {
```
I'm missing a few bits on SAML flow, so this might be an irrelevant comment; is it expected there's no 200 response ?

psauvage0 commented on selfservice/strategy/saml/strategy/strategy.go:
```diff
@@ -0,0 +1,352 @@
+			}
+
+			rf.UI.Nodes = append(rf.UI.Nodes, traitNodes...)
+			rf.UI.UpdateNodeValuesFromJSON(traits, "traits", node.OpenIDConnectGroup)
```
Here too

Stoakes commented on .schema/openapi.json:
```diff
@@ -2006,6 +2006,81 @@
     },
+    "/self-service/saml/idp":{
+      "get":{
+        "description": "This endpoint is to redirect the user to the idp auth flow",
```
All other descriptions are something like "This endpoint + verb at the active form".

psauvage0 commented on selfservice/strategy/saml/strategy/strategy.go:
```diff
@@ -0,0 +1,352 @@
+					continue
+				}
+
+				if parts[0] == conf.Providers[0].Provider && parts[1] == conf.Providers[0].Subject && len(conf.Providers[0].Subject) > 1 && len(conf.Providers[0].Provider) > 1 {
```
Possible OOB with `conf.Providers[0]` ?

Stoakes commented on embedx/config.schema.json:
```diff
@@ -312,6 +312,225 @@
+        },
+        "mapper_url": {
+          "title": "Jsonnet Mapper URL",
+          "description": "The URL where the jsonnet source is located for mapping the provider's data to Ory Kratos data.",
```
misc english wording

```suggestion
          "description": "Location for the jsonnet mapping between provider's data and Ory Kratos data",
```

Stoakes commented on embedx/config.schema.json:
```diff
@@ -312,6 +312,225 @@
+            },
+            "firstname": {
+              "title": "Firstname",
+              "description": "The name of the IDP attribute that will represent the user's firstname on Kratos",
```
Here and for other attributes

```suggestion
              "description": "Name of the IDP attribute representing user firstname on Kratos",
```

psauvage0 commented on selfservice/strategy/saml/strategy/strategy.go:
```diff
@@ -0,0 +1,352 @@
+func (s *Strategy) CountActiveFirstFactorCredentials(cc map[identity.CredentialsType]identity.Credentials) (count int, err error) {
+	for _, c := range cc {
+		if c.Type == s.ID() && gjson.ValidBytes(c.Config) {
+			var conf identity.CredentialsOIDC
```
`identity.CredentialsOIDC` in SAML strategy ?

Stoakes commented on persistence/sql/migrations/legacy/20220415152108_credentials_types_saml.up.fizz:
```diff
@@ -0,0 +1 @@
+sql("INSERT INTO identity_credential_types (id, name) SELECT 'ff5a1823-8b47-4255-860f-4b70ed122740', 'saml' WHERE NOT EXISTS ( SELECT * FROM identity_credential_types WHERE name = 'saml')")
```
Sorry for missing context: what is behind `ff5a1823-8b47-4255-860f-4b70ed122740` ?

psauvage0 commented on selfservice/strategy/saml/strategy/strategy.go:
```diff
@@ -0,0 +1,352 @@
+	return
+}
+
+func (s *Strategy) CountActiveMultiFactorCredentials(cc map[identity.CredentialsType]identity.Credentials) (count int, err error) {
```
Not implemented ?

sebferrer replied:
Not yet. This implementation isn't required since a long time, so we had to do it like this on our last rebase.

Stoakes commented on postgres.yaml:
```diff
@@ -0,0 +1,33 @@
+version: '3.7'
```
Is this file committed on purpose ?

Stoakes commented on postgres.yaml:
```diff
@@ -0,0 +1,33 @@
+
+services:
+  postgresd:
+      image: postgres:9.6
```
This postgres version is antic btw

Stoakes commented on internal/httpclient/model_self_service_saml_url.go:
```diff
@@ -0,0 +1,138 @@
+
+func (o SelfServiceSamlUrl) MarshalJSON() ([]byte, error) {
+	toSerialize := map[string]interface{}{}
+	if true {
```
This condition will always be true, so it can be removed.

Same for the other ones below.

Stoakes commented on internal/httpclient/model_self_service_saml_url.go:
```diff
@@ -0,0 +1,138 @@
+// GetSamlAcsUrl returns the SamlAcsUrl field value
+func (o *SelfServiceSamlUrl) GetSamlAcsUrl() string {
+	if o == nil {
+		var ret string
+		return ret
```
Could be compacted into a single `return ""` statement.

Stoakes commented on internal/httpclient/model_self_service_saml_url.go:
```diff
@@ -0,0 +1,138 @@
+// GetSamlMetadataUrl returns the SamlMetadataUrl field value
+func (o *SelfServiceSamlUrl) GetSamlMetadataUrl() string {
+	if o == nil {
+		var ret string
+		return ret
```
Same comment than L49

psauvage0 commented on selfservice/strategy/saml/strategy/strategy_auth.go:
```diff
@@ -0,0 +1,61 @@
+			return nil, err
+		}
+	} else {
+		// The user already exist in database so we log him
```
This makes me think : what happens when the IDP deletes a user ? I guess the user remains in the SP forever... What are the consequences of that ? It should not be a problem at all, but we could investigate this case a bit

ThibHrrd replied:
This is indeed what will happen but it is complicated for the SP to know that the IDP has deleted a user

Stoakes commented on selfservice/strategy/saml/provider_config.go:
```diff
@@ -0,0 +1,34 @@
+	// Label represents an optional label which can be used in the UI generation.
+	Label string `json:"label"`
+
+	// Represent the path of the certificate of your application
```
Standard Golang comment format is // <variable name> <description>

```suggestion
	// PublicCertPath path of your application certificate
```

Same for other variables below

Stoakes commented on selfservice/strategy/saml/provider_saml.go:
```diff
@@ -0,0 +1,62 @@
+
+	claims := &Claims{
+		Issuer:        "saml",
+		Subject:       SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["id"]),
```
Are we sure at this point of the code that `len(c.SAMLProviders)` is > 0 ?

Stoakes commented on selfservice/strategy/saml/provider_saml.go:
```diff
@@ -0,0 +1,62 @@
+		Picture:       SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["picture"]),
+		Email:         SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["email"]),
+		PhoneNumber:   SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["phone_number"]),
+		EmailVerified: true,
```
is `EmailVerified` truly always true at this point of the code ?

Stoakes commented on selfservice/strategy/saml/strategy/strategy.go:
```diff
@@ -0,0 +1,352 @@
+
+	// does not need sorting because there is only one field
+	c.SetCSRF(s.d.GenerateCSRFToken(r))
+	// AddSamlProviders(c, conf.Providers, message)
```
Commented code can be removed

Stoakes commented on selfservice/strategy/saml/strategy/strategy.go:
```diff
@@ -0,0 +1,352 @@
+	return err
+}
+
+func (s *Strategy) CountActiveCredentials(cc map[identity.CredentialsType]identity.Credentials) (count int, err error) {
```
Personal preference: returning explicitly values instead of using named return argument. I do so, because I find it less error prone.

Stoakes commented on selfservice/strategy/saml/strategy/strategy.go:
```diff
@@ -0,0 +1,352 @@
+	return
+}
+
+func (s *Strategy) CountActiveFirstFactorCredentials(cc map[identity.CredentialsType]identity.Credentials) (count int, err error) {
```
This function requires a bit of reading to fully understand what it does; it might be interesting to add a comment describing what it does.

Stoakes commented on selfservice/strategy/saml/strategy/strategy_auth.go:
```diff
@@ -0,0 +1,61 @@
+// This file is called when the /ACS receives an assertion, the method below allows you to define whether to register or login the user indicated in the assertion
+//#################
+
+// Handle SAML Assertion and process to either login or register
```
golang standard comment format starts with variable of function name.

Kratos uses a different one: https://github.com/ovh/kratos/blob/af1df63c62a8aec52f6adf9f23bb5dbd38c491ed/selfservice/flow/recovery/handler.go#L203

To be adapted based on your knowledge of ory coding style.

Stoakes commented on selfservice/strategy/saml/strategy/strategy_auth.go:
```diff
@@ -0,0 +1,61 @@
+			if err = s.processRegistration(w, r, registerFlow, provider, claims); err != nil {
+				if i == nil {
+					return nil, s.handleError(w, r, registerFlow, provider.Config().ID, nil, err)
+				} else {
```
Unused else: No need to add a `else` when last statement of an `if` is a return.

Here and below

psauvage0 commented on selfservice/strategy/saml/strategy/strategy_login.go:
```diff
@@ -0,0 +1,67 @@
+	return nil, nil
+}
+
+// Method not used but necessary to implement the interface
```
I find this weird but if you say so 🤨

sebferrer replied:
Here's the kind of error we get otherwise:
```
# github.com/ory/kratos/selfservice/strategy/saml/strategy
selfservice/strategy/saml/strategy/strategy_login.go:25:24: cannot use new(Strategy) (value of type *Strategy) as type login.Strategy in variable declaration:
        *Strategy does not implement login.Strategy (missing Login method)
```

psauvage0 replied:
What I found weird was that the method is called "Login", so I would expect that it is a pretty important method to implement in an IAM. 

Edit: As we discussed, it's because of the rebase against the origin. It's a new method added to the interface, probably worth checking where this method is called, to ensure that it is safe to not implement it

Stoakes reviewed (commented):
Nice work !

Some questions and a few comments. Mostly on English, Golang & code style.

psauvage0 commented on selfservice/strategy/saml/strategy/test/strategy_test.go:
```diff
@@ -0,0 +1,191 @@
+	is "gotest.tools/assert/cmp"
+)
+
+func TestGetAndDecryptAssertion(t *testing.T) {
```
Unrelated to the code : it would be interesting to discuss test strategy with the Ory community. It seems there is not a lot of info on what is considered a good level of testing on `CONTRIBUING.md` file (should we cover all branches of a functions ? Should we test for edge cases ? Or simply testing the main intent of the function is OK ?)

psauvage0 commented on spec/api.json:
```diff
@@ -3636,6 +3653,48 @@
         ]
       }
     },
+    "/self-service/methods/saml/auth": {
```
Same thing as before about the route : `/browser` vs `/api` routes

psauvage0 reviewed (changes requested):
Great work ! :D 

The main problem I see is with the attribute mapping that only works with a fixed list of attributes names. 

Otherwise it's about small code quality improvements. There are some pointed out by the linter that I didn't bother to comment as they are already shown in the PR builds.

I honestly think that we could already take reviews from the Ory team & community, as there are maybe some things about how they want the code to be organized.

aeneasr committed:
chore: format

aeneasr committed:
test(e2e): use correct dir

aeneasr committed:
test(e2e): improve webauthn test reliability

aeneasr committed:
fix: make hydra consistently localhost

aeneasr committed:
fix: panic

aeneasr committed:
fix: metadata should not be required

aeneasr committed:
ci: bump hydra

aeneasr committed:
test: fix oidc test regression

erolkskn committed:
fix: duplicate messages in recovery flow (#2592)

Urbansson committed:
fix: return browser to 'return_to' when logging in without registered account using oidc.  (#2496)

Close #2444

Co-authored-by: aeneasr <3372410+aeneasr@users.noreply.github.com>

jonas-jonas committed:
feat: add PATCH to adminUpdateIdentity (#2380) (#2471)

Closes #2380

Co-authored-by: aeneasr <3372410+aeneasr@users.noreply.github.com>

aeneasr committed:
refactor(sdk): rename `getJsonSchema` to `getIdentitySchema` (#2606)

BREAKING CHANGE: SDK Method `getJsonSchema` was renamed to `getIdentitySchema`.

ory-bot committed:
fix: format

ory-bot committed:
ci: add issues and PRs to board

CaptainStandby committed:
fix: re-issue outdated cookie in /whoami (#2598)

Closes #2562

Co-authored-by: aeneasr <3372410+aeneasr@users.noreply.github.com>
Co-authored-by: ory-bot <60093411+ory-bot@users.noreply.github.com>

aeneasr committed:
ci: fix version

aeneasr committed:
test: parallelize and speed up config tests (#2611)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

harnash committed:
feat: implement blocking webhooks (#1585)

feat: implement blocking webhooks (#1585)

This feature allows webhooks to return validation errors in the registration and login flow from a webhook. This feature enables you to deny sign-ups from a specific domain, for example.

A big thank you goes out to the team at Wikia / Fandom for implementing and contributing to this feature!

Closes #1724
Closes #1483

aeneasr committed:
ci: update project action

aeneasr committed:
ci: ignore schema YAMLs

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

erolkskn committed:
fix: disappearing title label on verification and recovery flow (#2613)

Closes #2591

t-tomalak committed:
fix: debugging Docker setup (#2616)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
fix(sdk): rust compile issue with required enum (#2619)

Co-authored-by: ory-bot <60093411+ory-bot@users.noreply.github.com>

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
chore: update repository templates

[skip ci] - updated repository templates to https://github.com/ory/meta/commit/32137ce422e5241bef5d7fd12c928ad39e34add0

aeneasr committed:
fix: revert removal of required field in uiNodeInputAttributes (#2623)

Co-authored-by: ory-bot <60093411+ory-bot@users.noreply.github.com>

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
feat: improve state generation logic

aeneasr committed:
fix: resolve bug where 500s in web hooks are not properly retried

aeneasr committed:
fix: respect more http sources for computing request URL

aeneasr committed:
fix: use full URL for webhook payload

aeneasr committed:
test: resolve builder regression

kevgo committed:
ci: delete semantic.yml (#2627)

kevgo committed:
ci: additional types and scopes for conventional commits (#2626)

aeneasr committed:
fix: cache migration status (#2631)

See https://github.com/ory-corp/cloud/issues/2691

aeneasr committed:
chore: update repository templates

[skip ci] - updated repository templates to https://github.com/ory/meta/commit/38af4400702ca1e7eeffe42270f5e9a23906bf63

aeneasr committed:
fix: remove jackc rewrites (#2634)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
fix: wrap migration error in WithStack (#2636)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
fix: improve migration status speed (#2637)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

jonas-jonas committed:
feat: replace message_ttl with static max retry count (#2638)

This PR replaces the `courier.message_ttl` configuration option with a `courier.message_retries` option to limit how often the sending of a message is retried before it is marked as `abandoned`. 

BREAKING CHANGES: This is a breaking change, as it removes the `courier.message_ttl` config key and replaces it with a counter `courier.message_retries`.

Closes https://github.com/ory/kratos/issues/402
Closes https://github.com/ory/kratos/issues/1598

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
refactor: hot reloading

sayoun committed:
fix(docker): add missing dependencies (#2643)

aeneasr committed:
chore: update repository templates

[skip ci] - updated repository templates to https://github.com/ory/meta/commit/93406389c87ece491fe9775500af1dd7e8aaa86f

aeneasr committed:
chore: update repository templates

[skip ci] - updated repository templates to https://github.com/ory/meta/commit/430183216bf2d03ee7e20d26e19000c2d55fbb54

aeneasr committed:
fix: migration error detection

jonas-jonas committed:
chore: fix golangci/lint version to v1.47.3

jonas-jonas committed:
fix: replace io/util usage by io and os package

aeneasr committed:
chore: update repository templates

[skip ci] - updated repository templates to https://github.com/ory/meta/commit/34c4f7ec6c0128421e0d055fd80ea21a6b27e78e

jonas-jonas committed:
chore: remove ioutil from open api templates

jonas-jonas committed:
fix: ignore CSRF for session extension on public route

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
fix: make servicelocator explicit

aeneasr committed:
Revert "autogen(openapi): regenerate swagger spec and internal client"

This reverts commit 4159b93ae3f8175cf7ccf77d34e4a7a2d0181d4f.

aeneasr committed:
fix: revert Go 1.19 formatting changes

aeneasr committed:
ci: fix to Go 1.18

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
fix: ignore error explicitly

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
fix: remove rust workaround

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
fix: re-add service to quickstart

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

vinckr committed:
docs: importing credentials supported

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

jonas-jonas committed:
fix: include flow id in use recovery token query (#2679)

This PR adds the `selfservice_recovery_flow_id` to the query used when "using" a token in the recovery flow.

This PR also adds a new enum field for `identity_recovery_tokens` to distinguish the two flows: admin versus self-service recovery.

BREAKING CHANGES: This patch invalidates recovery flows initiated using the Admin API. Please re-generate any admin-generated recovery flows and tokens.

aeneasr committed:
fix: ignore commata in HIBP response

aeneasr committed:
fix: make swagger/openapi go 1.19 compatible

aeneasr committed:
feat: support ip exceptions

aeneasr committed:
chore: go 1.19 format

aeneasr committed:
feat: add missing config entry

ThibHrrd committed:
feat(saml): saml 2.0 implementation
Signed-off-by: ThibaultHerard <thibaultherard10@gmail.com>

Co-authored-by: sebferrer <sebferrer@users.noreply.github.com>
Co-authored-by: psauvage <psauvage0@users.noreply.github.com>
Co-authored-by: alexGNX <alexGNX@users.noreply.github.com>
Co-authored-by: Stoakes <Stoakes@users.noreply.github.com>

aeneasr committed:
feat: add pre-hooks to settings, verification, recovery

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

hperl committed:
fix: do not invalidate recovery addr on update (#2699)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

kevgo committed:
chore: improve package-lock.json package.json (#2712)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

jonas-jonas committed:
feat: add flow id check to use verification token (#2695)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

LandonPattison committed:
feat: allow importing scrypt hashing algorithm (#2689)

It is now possible to import scrypt-hashed passwords.

See #2422

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

jonas-jonas committed:
fix: patch invalidates credentials (#2721)

Fixes ory/cloud#148

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
feat: add missing cookie headers to SDK methods (#2720)

See https://github.com/ory/kratos/discussions/2583

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

floriankramer committed:
fix: check return code of ms graphapi /me request. (#2647)

Co-authored-by: Jonas Hungershausen <jonas.hungershausen@ory.sh>

erolkskn committed:
feat: add api endpoint to fetch messages (#2651)

Closes https://github.com/ory/kratos/issues/2639

Co-authored-by: aeneasr <3372410+aeneasr@users.noreply.github.com>

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
chore: update repository templates

[skip ci] - updated repository templates to https://github.com/ory/meta/commit/ac6bef904ee4d135b9d5ee85418716237cf09df0

aeneasr committed:
fix: remove jsonnet import support

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

kevgo committed:
chore: format using Make (#2736)

kelkarajay committed:
fix: format check stage in the CI (#2737)

kevgo committed:
Remove empty script (#2739)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/935cc0443464fd76fbf41dff1081b368080c9353

kevgo committed:
chore: update ory-prettier-styles (#2749)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/4ef13422e91f15b9f70014a0d67b92498ab728d1

kevgo committed:
chore: fix formatting (#2753)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

kevgo committed:
chore: remove double-tabs in Makefile (#2747)

kevgo committed:
chore: format using Make (#2748)

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/19eed817e5d5b64509887ef5f1e3eff3e3ce03a1

kevgo committed:
chore: remove listx dependency (#2752)

kelkarajay committed:
chore: remove dead code (#2769)

jwmay2012 committed:
feat: add support for firebase scrypt hashes on identity import and login hash upgrade (#2734)

See #2422

zepatrik committed:
refactor: use gotemplates for command usage (#2770)

aeneasr committed:
Revert "refactor: use gotemplates for command usage (#2770)" (#2778)

This reverts commit 1d22b235291ce7102dd186a53a431b55780973d3.

zepatrik committed:
ci: remove deprecated linters (replaced by `unused`)

zepatrik committed:
refactor: use gotemplates for command usage

zepatrik committed:
chore: bump go to 1.19

alnr committed:
feat: automatic TLS certificate reloading (#2744)

jonas-jonas committed:
chore: add node version check to test/e2e/run.sh (#2745)

Resolves #2738

mmeller-wikia committed:
fix: preserve return_to param between flows (#2644)

kelkarajay committed:
feat: adding device information to the session (#2715)

Closes https://github.com/ory/kratos/issues/2091
See https://github.com/ory-corp/cloud/issues/3011

Co-authored-by: hackerman <3372410+aeneasr@users.noreply.github.com>
Co-authored-by: Patrik <zepatrik@users.noreply.github.com>

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
chore: update x/sys for M1

pr1ze committed:
feat: parse all id token claims into raw_claims (#2765)

All ID Token claims resulting from the Social Sign In flow are now available in `raw_claims` and can be used in the Social Sign In JsonNet Mapper.

Closes #2528

aeneasr committed:
fix: proper annotation for patch (#2784)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

jonas-jonas committed:
feat: replace magic links with one time codes in recovery flow (#2645)

This feature introduces a new `code` strategy to recover an account. 

Currently, if a user needs to initiate a recovery flow to recover a lost password/MFA/etc., they’ll receive an email containing a “magic link”. This link contains a flow_id and a recovery_token. This is problematic because some antivirus software opens links in emails to check for malicious content, etc.

Instead of the magic link, we send an 8-digit code that is clearly displayed in the email or SMS. A user can now copy/paste or type it manually into the text-field that is shown after the user clicks “submit” on the initiate flow page.

Closes #1451

BREAKING CHANGES: This patch changes the behavior of the recovery flow. It introduces a new strategy for account recovery that sends out short "one-time passwords" (`code`) that a user can use to prove ownership of their account and recovery access to it. This PR also updates the default recovery strategy to `code`.

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/23d918a32533554c30d720dc44e821de3cda18f8

brahmlower committed:
fix: include metadata_admin in admin identity list response (#2791)

Closes #2711

Commit:
docs: cypress on arm based mac (#2795)

brahmlower committed:
fix: identity sessions list response includes pagination headers (#2763)

Closes #2762

Benehiko committed:
fix: express e2e tests for new account experience (#2708)

jonas-jonas committed:
fix: missing usage to recovery_code_invalid template (#2798)

jonas-jonas committed:
fix: not cleared field validation message (#2800)

Commit:
feat: immutable cookie session values (#2761)

Closes #2701

Commit:
fix: x-forwarded-for header parsing (#2807)

aeneasr committed:
feat: add cache headers (#2817)

aeneasr committed:
feat: improve cache handling

kmherrmann committed:
feat: update for the Ory Network (#2814)

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/6ab5ce6da0cc57d4492e932602bbfd4a76547795

jonas-jonas committed:
fix: correct name of span on recovery code deletion (#2823)

bplcensys committed:
docs: cleanup v0alpha2 endpoint summaries

bplcensys committed:
fix(lint): fixed lint error causing ci failures

zepatrik committed:
fix: correctly calculate `expired_at` timestamp for FlowExpired errors (#2836)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
feat: OAuth2 integration (#2804)

This feature allows Ory Kratos to act as a login provider for Ory Hydra using the `oauth2_provider.url` configuration value.

Closes https://github.com/ory/kratos/issues/273
Closes https://github.com/ory/kratos/discussions/2293
See https://github.com/ory/kratos-selfservice-ui-node/pull/50
See https://github.com/ory/kratos-selfservice-ui-node/pull/68
See https://github.com/ory/kratos-selfservice-ui-node/pull/108
See https://github.com/ory/kratos-selfservice-ui-node/pull/111
See https://github.com/ory/kratos-selfservice-ui-node/pull/149
See https://github.com/ory/kratos-selfservice-ui-node/pull/170
See https://github.com/ory/kratos-selfservice-ui-node/pull/198
See https://github.com/ory/kratos-selfservice-ui-node/pull/207

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

mmeller-wikia committed:
docs: enable 2FA methods in docker-compose quickstart setup (#2828)

mmeller-wikia committed:
fix: take over return_to param from unauthorized settings to login flow (#2787)

jonas-jonas committed:
fix: wrong config key in admin recovery documentation (#2815)

aeneasr committed:
chore: debugf (#2842)

aeneasr committed:
chore(sdk): update order of arguments (#2840)

aeneasr committed:
fix(sdk): identity metadata is nullable (#2841)

Closes https://github.com/ory/sdk/issues/218

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/4a68ca0e3b70305c4a49a65777cb7f83a5eb9d89

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

Commit:
feat: adding admin session listing api (#2818)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

jonas-jonas committed:
fix: admin recovery CSRF & duplicate form elements (#2846)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/852a1aece5fefac0a03f928672538c5d8c536ad8

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/b41b1ee5ed62f47bac563014929c64bff0c14163

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/a2fba7e968572391ac4a55ce362dca0c4800c97d

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/d3f8710e356fb833d4bd71b4ba19d062df2ea89e

kevgo committed:
chore: license checker (#2851)

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/9f57fecccae6e37a4b7ff5863a683d27e583cd0f

kevgo committed:
chore: remove obsolete header (#2857)

kevgo committed:
chore: fix package-lock.json (#2861)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
refactor: SDK v1 naming

Find the full [upgrade guide in our documentation](https://www.ory.sh/docs/guides/upgrade/sdk).

Commit:
feat: add admin get api for session (#2855)

Co-authored-by: hackerman <3372410+aeneasr@users.noreply.github.com>

alnr committed:
feat: add OpenTelemetry span events (#2858)

Co-authored-by: Ajay Kelkar <ajay.kelkar@ory.sh>

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/47569d9893f0bae29676417807de790338fec9be

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

jonas-jonas committed:
chore: update ory/x (#2871)

aeneasr committed:
fix: use process-isolated Jsonnet VM (#2869)

Co-authored-by: hperl <34397+hperl@users.noreply.github.com>

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

zepatrik committed:
fix: distinguish credential types properly when collecting identifiers (#2873)

jonas-jonas committed:
fix: make `courier.TemplateType` an enum (#2875)

Commit:
feat: add handler with openapi def for admin revoke session (#2867)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
refactor: make embedding easier with internal sdk

kevgo committed:
feat: standardize license headers (#2790)

kevgo committed:
test: update snapshots (#2877)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
fix: codecov (#2879)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
feat: upgrade hydra to v2

aeneasr committed:
ci: add moreutils and gettext

aeneasr committed:
chore: format

aeneasr committed:
test(e2e): fix issuer config

alnr committed:
chore: bump ory/x (#2882)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
feat: ingest hydra bugfix

kevgo committed:
chore: consolidate .gitignore files (#2881)

kevgo committed:
chore: list contributors in file (#2878)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
chore: add additional files to gitignore

aeneasr committed:
test: improve e2e tooling

aeneasr committed:
ci: update hydra

aeneasr committed:
test(e2e): resolve max-age regression

aeneasr committed:
test(e2e): resolve max-age and issuer regression

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
fix: do not double-commit webhooks on registration (#2888)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
fix: do not crash process on invalid smtp url (#2890)

Closes https://github.com/ory-corp/cloud/issues/3321

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

jonas-jonas committed:
feat: add verification via `code` (#2838)

The new `code` strategy is now supported as a verification strategy. If enabled, the strategy sends a code, instead of a magic link to the user's address, which they can use to verify their address.

Close #2824

aeneasr committed:
fix: incorrect swagger annotation for `getSession` (#2891)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

jonas-jonas committed:
chore: remove .only from test (#2893)

ory-bot committed:
fix: add missing go.mod to docker build

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

Commit:
feat: add session cache header feature flag (#2899)

See ory-corp/cloud#3283

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

jonas-jonas committed:
feat: change code length to 6 numbers (#2894)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

Commit:
fix: active attribute based off IsActive checks (#2901)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

jonas-jonas committed:
feat: add "success" UITextType (#2900)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

jonas-jonas committed:
fix: verification redirect & continue label (#2905)

This PR resolves an issue with the redirect after a successful verification, if not specified.

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

Commit:
fix: set cache default to false (#2906)

jonas-jonas committed:
fix: send out correct verification invalid email in code strategy (#2908)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

erolkskn committed:
feat: support md5 hash import (#2725)

ThibHrrd committed:
feat(saml): use ory/x fetcher
Signed-off-by: ThibaultHerard <thibaultherard10@gmail.com>

Co-authored-by: sebferrer <sebferrer@users.noreply.github.com>

ThibHrrd committed:
feat(saml): improved error handling
Signed-off-by: ThibaultHerard <thibaultherard10@gmail.com>

Co-authored-by: sebferrer <sebferrer@users.noreply.github.com>

martinei committed:
feat: trace WebHooks (#2911)

Previously the context was not propagated to the http client. As a result the (instrumented) client did not find the existing span and the sapns for outgoing http request have been orphains.

With this simple Fix they are now children of the corresponding webhook spans.

ory-bot committed:
autogen: pin v0.11.0-alpha.0.pre.2 release commit

alnr committed:
feat: forward parsed request cookies to webhook Jsonnet snippet (#2917)

Request cookies were already available in raw form in
the ctx.request_headers top-level argument to the Jsonnet snippet.
Parsing cookies in Jsonnet is tedious and error-prone, though, so
we parse them internally for convenience.

ory-bot committed:
autogen: pin v0.11.0 release commit

ory-bot committed:
autogen(docs): generate and bump docs

[skip ci]

ory-bot committed:
autogen: add v0.11.0 to version.schema.json

[skip ci]

aeneasr committed:
feat: require verification on login (#2927)

jonas-jonas committed:
feat: store errors of courier message (#2914)

BREAKING CHANGE: The `/admin/courier/messages` endpoint now uses `keysetpagination` instead.

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

kstasik committed:
fix: cors headers not added to the response #2922 (#2934)

sebferrer committed:
feat(saml): relaystate continuity fix + unit tests
Signed-off-by: sebferrer <sebferrer@users.noreply.github.com>

Co-authored-by: ThibaultHerard <thibaultherard10@gmail.com>
